judge/judgedaemon.main.php

```diff
@@ -290,6 +290,39 @@ function read_judgehostlog(int $numLines = 20) : string
     return trim(ob_get_clean());
 }
 
+/**
+ * Execute a shell command given an array of strings forming the command.
+ * The command and all its arguments are automatically escaped.
+ *
+ * @param array    $command_parts         The command and its arguments (e.g., ['ls', '-l', '/tmp/']).
+ * @param int|null $retval                The variable to store the command's exit code.
+ * @param bool     $log_nonzero_exitcode  Whether non-zero exit codes should be logged.
+ *
+ * @return bool Returns true on success (exit code 0), false otherwise.
+ */
+function run_command_safe(array $command_parts, ?int &$retval = null, $log_nonzero_exitcode = true): bool
```

Comment:
When I look at most call sites, then `$retval` is not used afterwards, so it would be a bit simpler to be able to not pass it and then only within this function a local `$retval` variable is used.

Reply:
I didn't realize you couldn't just use `null`, but if this works, LGTM!